didcomm_messaging/routing.py

```diff
@@ -47,8 +49,70 @@ async def prepare_forward(
             encoded_message (bytes): The encoded message.
 
         Returns:
-            The encoded message, and the service endpoint to forward to.
+            The encoded message, and the services to forward to.
         """
-        service = await self._resolve_service(to)
-        # TODO Do the stuff
-        return encoded_message, service
+
+        # Get the initial service
+        services = await self._resolve_services(to)
+        chain = [
+            {
+                "did": to,
+                "service": services,
+            }
+        ]
+
+        # Loop through service DIDs until we run out of DIDs to forward to
+        to_did = services[0].service_endpoint.uri
+        found_forwardable_service = await self.is_forwardable_service(services[0])
+        while found_forwardable_service:
+            services = await self._resolve_services(to_did)
+            if services:
+                chain.append(
+                    {
+                        "did": to_did,
+                        "service": services,
+                    }
+                )
+                to_did = services[0].service_endpoint.uri
+            found_forwardable_service = (
+                await self.is_forwardable_service(services[0]) if services else False
+            )
+
+        if not chain[-1]["service"]:
+            raise RoutingServiceError(f"No DIDCommV2 service endpoint found for {to}")
+
+        # Grab our target to pack the initial message to, then pack the message
+        # for the DID target
+        next_target = chain.pop(0)
+        packed_message = encoded_message
+
+        # Loop through the entire services chain and pack the message for each
+        # layer of mediators
+        for service in chain:
+            packed_message = await self.packaging.pack(
+                json.dumps(
+                    {
+                        "typ": "application/didcomm-plain+json",
+                        "type": "https://didcomm.org/routing/2.0/forward",
+                        "id": str(uuid.uuid4()),
+                        # "frm": did, # Do we need this? Spec shows it missing in the example
+                        "to": [service["did"]],
+                        "body": {"next": next_target["did"]},
+                        "attachments": [
+                            {
+                                "id": str(uuid.uuid4()),
+                                "media_type": "application/didcomm-encrypted+json",
+                                "data": {
+                                    "json": json.loads(packed_message),
+                                },
+                            },
+                        ],
+                    }
+                ),
+                [service["did"]],
+            )
```

Comment:
For now, I've split this out into a helper function. I would love to see a "forward settings" thing get passed down that allows us to expire forwarded messages (since it is part of the spec)